geode-core/src/main/java/org/apache/geode/internal/cache/DistributedRegion.java

```diff
@@ -526,21 +526,36 @@ public boolean hasSeenEvent(EntryEventImpl event) {
     isDuplicate = getEventTracker().hasSeenEvent(event);
     if (isDuplicate) {
       markEventAsDuplicate(event);
+      if (event.getVersionTag() == null && event.getRegion().getConcurrencyChecksEnabled()
```

Comment:
We could move all this checks to findAndSetVersionTag()
Then, it will be:

public boolean hasSeenEvent(EntryEventImpl event) {
    boolean isDuplicate;
    isDuplicate = getEventTracker().hasSeenEvent(event);
    if (isDuplicate) {
      markEventAsDuplicate(event);
    } 
    if (event.isPossibleDuplicate()) {
      findAndSetVersionTag(event);
    }
    return isDuplicate;
  }

  void findAndSetVersionTag(EntryEventImpl event) {
    if (event.isPossibleDuplicate() && event.getRegion().getConcurrencyChecksEnabled()
        && event.getVersionTag() == null && event.getEventId() != null) {
      boolean isBulkOp = event.getOperation().isPutAll() || event.getOperation().isRemoveAll();
      VersionTag tag =
          FindVersionTagOperation.findVersionTag(event.getRegion(), event.getEventId(), isBulkOp);
      if (tag == null) {
        logger.info("No version tag can be found in cluster when retrying the following event {}",
            event);
      }
      event.setVersionTag(tag);
    }
  }